include/hdf5/group.h

```diff
@@ -0,0 +1,65 @@
+//
+// ICRAR - International Centre for Radio Astronomy Research
+// (c) UWA - The University of Western Australia, 2017
+// Copyright by UWA (in the framework of the ICRAR)
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+/**
+ * @file
+ *
+ * C++ wrappers for dealing HDF5 groups (abstract or concrete)
+ */
+
+#ifndef SHARK_HDF5_GROUP_H
+#define SHARK_HDF5_GROUP_H
+
+#include "hdf5/location.h"
+
+namespace shark {
+namespace hdf5 {
+
+class Group;
+
+class DataSet;
+
+class AbstractGroup : public Location {
+public:
+	AbstractGroup(H5I_type_t expectedType, hid_t handle);
+
+	hsize_t getNumObjs() const;
+	std::string getObjnameByIdx(hsize_t idx) const;
+	H5G_obj_t getObjTypeByIdx(hsize_t idx) const;
+	Group openGroup(const std::string& name) const;
+	DataSet openDataSet(const std::string& name) const;
+	Group createGroup(const std::string& name);
+	DataSet createDataSet(const std::string& name, const DataType& dataType, const DataSpace& dataSpace);
+};
+
+class Group : public AbstractGroup {
+public:
+	Group(const AbstractGroup& parent, const std::string& name);
+	~Group() override;
+
+	static Group create(AbstractGroup& parent, const std::string& name);
```

Comment:
I've opted in most cases (only exception is `File` class) to have a separate `create()` static member function. Reason being that in many cases the signature is identical to what is required just to open the resource. A down side of this is that we have to add a private constructor that takes the created handle/id.

An alternative would be to have some sort of enum to control `open`/`create` logic in the constructor in the case where the signatures are identical (and an overload in the case where they are different). I kind of like the static constructor (Java influences there I think :wink: ), but also the private constructor is also a little yucky & can see the case for both sides.

Thoughts?

Reply:
That's fine, either way would create a bit more of trouble in one direction or the other, let's stick with the static `create` functions.